_drafts/arm-cortex-m-exceptions-and-nvic.md

```diff
@@ -0,0 +1,699 @@
+- **Inactive** - The exception is neither pending or active
+
+Some exceptions can be selectively Enabled or Disabled.
+
+> NOTE: Even while an exception is disabled, it can still reach the pending state. Upon being
+> enabled it will then transition to active. It's generally a good idea to clear any pending exceptions for an interrupt before enabling them.
+
+Let's explore the different types of exceptions available on ARM Cortex-M MCUs:
+
+### Built in Exceptions
+
+These are exceptions that are part of every ARM Cortex-M core. The ARM Cortex-M specifications
+reserve **Exception Number**s **1**-**15**, inclusive, for these.
+
+> NOTE: Recall that the **Exception Number** maps to an offset within the **Vector Table**. Index 0 of
+> the Vector Table holds the reset value of the Main stack pointer. The rest of the **Vector Table**
+> holds **Exception Handler** pointers which is why the **Exception Number**s start at 1
+
+Six exception handlers are always supported and depending on the Cortex-M variant, additional
+handlers will be implemented as well. The minimum set is:
+
+- **Reset** - This is the routine executed when a chip comes out of reset. More details can be found within the zero-to-main bootloader series of posts[^2]
+- **Non Maskable Interrupt** (`NMI` ) - As the name implies, the interrupt cannot be disabled. If
+  errors happen in other exception handlers, a NMI will be triggered. Aside from the `Reset`
+  exception, it has the highest priority of all exceptions.
+- **HardFault** - The catchall for assorted system failures that can take place such as accesses to
+  bad memory, divide-by-zero errors and illegal unaligned accesses. It's the only handler for
+  faults on the ARMv6-M architecture but for ARMv7-M & ARMv8-M, finer granularity fault handlers
+  can be enabled for specific error classes (i.e `MemManage`, `BusFault`, `UsageFault`) [^3]
+- **SVCall** - Exception handler invoked when an _Supervisor Call_ (`svc`) instruction is executed
+- **PendSV** & **SysTick** - System level interrupts triggered by software. They are typically used
+  when running a RTOS to manage when the scheduler runs and when context switches take place.
+
+### External Interrupts
+
+ARM cores also support interrupt lines which are "external" to the core itself. These interrupt
+lines are usually routed to vendor-specific peripherals on the MCU such as Direct Memory Access (**DMA**)
+engines or General purpose input/output pins (**GPIO**s). All of these interrupts are configured via a peripheral known as the _Nested Vectored Interrupt Controller_ (**NVIC**).
+
+The **Exception Number** for external interrupts starts at **16**. The ARMv7-M reference manual has a good graphic which displays the Exception number mappings:
+
+![](img/armv-m-exceptions/exception-numbers.png)
+
+## Registers used to configure Cortex-M Exceptions
+
```

Comment:
I would add a note here to allow folks to skip the section. Something like "If you're already familiar with NVIC register, click here to go straight to examples".

Reply:
Great idea ... I added:

> If you already have a good feel for Cortex-M exception configuration, I'd recommend skipping to the
> [advanced topics section](#advanced-topics) which covers a couple different implementation
> details worth noting when dealing with ARM interrupts or test your knowledge with a more [complex configuration example](#complex-config-example)!

I'm now leaning towards keeping the advanced stuff in and just re-directing readers